keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,145 @@
+---
+
+# Add maxInitialFailureCount to health probes
+
+## Table of Contents
+
+- [Add maxInitialFailureCount to health probes](#add-maxinitialfailurecount-to-health-probes)
+  - [Table of Contents](#table-of-contents)
+  - [Release Signoff Checklist](#release-signoff-checklist)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Implementation Details](#implementation-details)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+  - [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Slow starting containers are difficult to address with the current status of health probes: they are either killed before being up, or could be left dead-locked during a very long time before being killed.
+
+This proposal adds a numerical option `maxInitialFailureCount` to probes allowing a greater number of failures during the initial start of the container before taking action, while keeping `failureThreshold` at a minimum to restart dead-locked containers after an acceptable delay.
+
+## Motivation
+
+The main problem with slow starting containers is that you have to leave them enough time to start before having `livenessProbe` fail `failureThreshold` times (otherwise `kubelet` kills them before they have a chance to be up).
+This is even worse when load on the nodes can increase this long startup time since you have to take into account the worse case scenario (let's call it **T**).
+
+You can handle that today with various strategies:
+
+- Delay the initial `livenessProbe` sufficiently to permit the container to start up (set `initialDelaySeconds` greater than **T**). While this ensures that the `livenessProbe` will not fail during the startup period, it also prevents the container from being killed early in case of a deadlock. Also, since the `livenessProbe` isn't run during that delay, there is no feedback loop on the container state.
```

Comment:
How does setting `initialDelaySeconds` prevent the container from being killed early if it deadlocks any more than this proposal does?  Is the concern about the container coming up (so that a liveness probe would see it succeed) and only after that deadlock?  If that's what you mean, say so.  If it's something else, please state that as well.